packaging/metadata.py

```diff
@@ -0,0 +1,416 @@
+    # 1.1
+    supported_platform: Collection[str] = ()
+    download_url: str = ""
+    classifier: Collection[str] = ()
+    # 1.2
+    maintainer: str = ""
+    maintainer_email: Collection[EmailAddress] = ()
+    requires_dist: Collection[Requirement] = ()
+    requires_python: SpecifierSet = dataclasses.field(default_factory=SpecifierSet)
+    requires_external: Collection[str] = ()
+    project_url: Mapping[str, str] = dataclasses.field(default_factory=dict)
+    provides_extra: Collection[str] = ()
+    provides_dist: Collection[Requirement] = ()
+    obsoletes_dist: Collection[Requirement] = ()
+    # 2.1
+    description_content_type: str = ""
+    # 2.2
+    dynamic: Collection[NormalizedDynamicFields] = ()
+
+    def __post_init__(self) -> None:
+        """Perform required data conversions, validations, and ensure immutability"""
+
+        _should_be_set = (self._MULTIPLE_USE | {"keywords"}) - {"project_url"}
+
+        for field in self._fields():
+            value = getattr(self, field)
+            if field.endswith("dist"):
+                reqs = (self._convert_single_req(v) for v in value)
+                _setattr(self, field, frozenset(reqs))
+            elif field.endswith("email"):
+                emails = self._convert_emails(value)
+                _setattr(self, field, frozenset(emails))
+            elif field in _should_be_set:
+                _setattr(self, field, frozenset(value))
+            elif field in {"description", "summary"}:
+                _setattr(self, field, value.strip())
+
+        urls = self.project_url
+        if not isinstance(urls, Mapping) and isinstance(urls, Iterable):
+            urls = {}
+            for url in cast(Iterable[str], self.project_url):
+                key, _, value = url.partition(",")
+                urls[key.strip()] = value.strip()
+        _setattr(self, "project_url", ImmutableMapping(urls))
+
+        # Dataclasses don't enforce data types at runtime
+        if not isinstance(self.requires_python, SpecifierSet):
+            requires_python = self._parse_requires_python(self.requires_python)
+            _setattr(self, "requires_python", requires_python)
+        if self.version and not isinstance(self.version, Version):
+            _setattr(self, "version", Version(self.version))
+
+        if self.dynamic:
+            values = (_normalize_field_name_for_dynamic(f) for f in self.dynamic)
+            dynamic = frozenset(v for v in values if self._validate_dynamic(v))
+            _setattr(self, "dynamic", dynamic)
+
+    @property
+    def metadata_version(self) -> str:
+        """
+        The data structure is always compatible with the latest approved
+        version of the spec, even when parsing files that use previous versions.
+        """
+        return "2.2"
+
+    @classmethod
+    @lru_cache(maxsize=None)
+    def _fields(cls) -> Collection[str]:
+        return frozenset(f.name for f in dataclasses.fields(cls))
```

Comment:
Is this so costly it warrants maintaining a cache?

Reply:
You are completely right, this is definitely a micro-optmisation.
I removed the cache in one of the following commits.